README.md
# `micro-github`

# SrBrahma fork:

My Typescript fork with major changes from another fork (https://github.com/zalo/micro-github-1) that updates the repository micro-github (https://github.com/mxstbr/micro-github). It OAuth'enticates GitHub users, retrieving the token to http://localhost:${clientChosenPort}/.

It was made / changed to work with my other project, the https://github.com/SrBrahma/GitHub-Repository-Manager.

Removed the original README as many stuff don't apply anymore. If you are interested in using my code, you should try first the (https://github.com/zalo/micro-github-1) and/or read its README as a knowledge base (or open a previous commit, linked below).


Since 2.0.0, it hardcodes the callback URL to the localhost and requires the client to send the port before. If you want to use a non-localhost callback URL, you may use this previous version / commit: https://github.com/SrBrahma/micro-github/tree/b2e3658b372d7c181665540d639254870b3056cb .


You may create an Issue asking for help or for an update ;)

Only leaving here the original Licence.


## License

Copyright (c) 2017 Maximilian Stoiber, licensed under the MIT license. See [LICENSE.md](LICENSE.md) for more information.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

# SrBrahma fork

## [2.0.0] - May 24, 2020

Removed allow_signup option. Now always true.

Now always redirects to localhost.

Client must provide the redirectPort.

Displays an error page on error, only redirects to user on success (as with a invalid redirectPort query param, it wouldn't have an port to send the error code).


## [?.?.?] - May 4, 2020

Hardcoded the scope parameter, so malicious applications that have access to the user machine can't mistake him under your application name to get further authorizations.

Added max states length, so malicious people can't explode Vercel RAM with invalid requests.


## [?.?.?] - April 30, 2020

Moved to typescript with the necessary changes.

Minor general changes.


# Parent Repository original Change Log

## [1.1.0] - 2019-08-07
### Added
- Refactor to support Zeit Now's 2.0 routing system.

## [1.0.0] - 2017-04-01
### Added
- Support for state parameter to protect against cross-site request forgery attacks. Requires users to use the provided `/login` url and moves redirect url from `/` to `/callback`


api/[action].js
const querystring = require('querystring');
const axios = require('axios');
const redirect = require('micro-redirect');
const uid = require('uid-promise');

const githubUrl = process.env.GH_HOST || 'github.com';

const states = [];

const redirectWithQueryString = (res, data) => {
  const location = `${process.env.REDIRECT_URL}?${querystring.stringify(data)}`;
  redirect(res, 302, location);
};

const login = async (req, res) => {
  const state = await uid(20);
  states.push(state);
  const { scope, allow_signup } = req.query;
  const query = {
    client_id: process.env.GH_CLIENT_ID,
    state: state
  };
  if (scope) query.scope = scope;
  if (allow_signup !== undefined) query.allow_signup = allow_signup;
  redirect(res, 302, `https://${githubUrl}/login/oauth/authorize?${querystring.stringify(query)}`);
};

const callback = async (req, res) => {

  res.setHeader('Content-Type', 'text/html');
  const { code, state } = req.query;

  if (!code && !state) {
    redirectWithQueryString(res, { error: 'Provide code and state query param' });
  } else if (!states.includes(state)) {
    redirectWithQueryString(res, { error: 'Unknown state' });
  } else {
    states.splice(states.indexOf(state), 1);
    try {
      const { status, data } = await axios({
        method: 'POST',
        url: `https://${githubUrl}/login/oauth/access_token`,
        responseType: 'json',
        data: {
          client_id: process.env.GH_CLIENT_ID,
          client_secret: process.env.GH_CLIENT_SECRET,
          code
        }
      });

      if (status === 200) {
        const qs = querystring.parse(data);
        if (qs.error) {
          redirectWithQueryString(res, { error: qs.error_description });
        } else {
          redirectWithQueryString(res, { access_token: qs.access_token });
        }
      } else {
        redirectWithQueryString(res, { error: 'GitHub server error.' });
      }
    } catch (err) {
      redirectWithQueryString(res, { error: 'Please provide GH_CLIENT_ID and GH_CLIENT_SECRET as environment variables. (or GitHub might be down)' });
    }
  }
};

// api/[action].js
// https://micro-github.*USERNAME*.now.sh/api/login
// https://micro-github.*USERNAME*.now.sh/api/callback
module.exports = (req, res) => {
  if (req.query.action === "login") {
    login(req, res);
  } else if (req.query.action === "callback") {
    callback(req, res);
  }
};


api/[action].ts
// You must set these env vars in your Vercel:
// GH_CLIENT_ID
// GH_CLIENT_SECRET

// TODO: Add a major version system for easier changes.
// TODO: Something like this in state:
// [random string]_p[port: number]_v[version: number]


require('dotenv').config(); // dotEnv used on development mode, getting the env vars from .env file.
import { NowRequest, NowResponse } from '@now/node';
import querystring from 'querystring';
import axios from 'axios';
import uid from 'uid-promise';

// https://developer.github.com/apps/building-oauth-apps/understanding-scopes-for-oauth-apps/
// "Note: Your OAuth App can request the scopes in the initial redirection. You can specify multiple
// scopes by separating them with a space". Ex: 'repo gist'
const scope = 'repo';


// state format is uidString_customRedirectPort.
// We store the redirect port with the state to use less Vercel computation power (to
// use indexOf instead of findIndex. It is way faster.)

// If someday this stops working, maybe wouldn't be a problem to always trust the received state.
// Don't see a security issue with this, at the current moment.

// For additional security, it could be KEY_CRYPTO,
// where the key would still be a random value, and crypto a value generated using the key and a Env secret.
// Maybe doesn't make sense, as the toGithub redirect contains the state, and the user could always use this
// state. Use a Date()?
const states: string[] = [];
const statesMaxLength = 10000; // A big number. May be changed.



function redirect(res: NowResponse, location: string, bodyObj?: any) {
  res.status(302);
  res.setHeader('Location', location + (bodyObj ? `?${querystring.stringify(bodyObj)}` : ''));
  res.send('');
}


function redirectToUser(res: NowResponse, bodyObj: any, port: number) {
  //TODO: After ~1 july, remove the '/oauthCallback'.
  // The user should already have the extension updated to allow callback to '/'.
  redirect(res, `http://localhost:${port}/oauthCallback`, bodyObj);
};



async function login(req: NowRequest, res: NowResponse) {
  let { redirectPort: redirectPortString } = req.query;

  //TODO: Remove this after ~1 july. Is just a "migration". And change above to const.
  if (!redirectPortString)
    redirectPortString = "60002";

  let redirectPort = Number(redirectPortString); // If undefined, returns NaN

  if (redirectPort < 0 || redirectPort > 65535 || !Number.isInteger(redirectPort)) // If NaN, isNotInteger.
    return errorPage(res, 'Login Error 1: Invalid or missing redirect port.');


  // To avoid the Vercel RAM exploding, if callbacks doesn't happen for a reason (aka evil users hehe)
  if (states.length >= statesMaxLength)
    states.shift();

  const state = `${await uid(20)}_${redirectPort}`;
  states.push(state);

  const query = {
    client_id: process.env.GH_CLIENT_ID,
    state,
    scope,
  } as any;

  redirect(res, `https://github.com/login/oauth/authorize`, query);
};

async function callback(req: NowRequest, res: NowResponse) {
  const { code, state } = req.query;

  if (Array.isArray(code) || Array.isArray(state) || !code || !state)
    return errorPage(res, 'Callback Error 1: Invalid query');

  // Find the state in states array
  const stateIndex = states.indexOf(state);

  if (stateIndex == -1)
    return errorPage(res, 'Callback Error 2: State not found');

  // Remove the state from states array
  states.splice(stateIndex, 1);

  try {
    const { status, data } = await axios({
      method: 'POST',
      url: `https://github.com/login/oauth/access_token`,
      responseType: 'json',
      data: {
        client_id: process.env.GH_CLIENT_ID,
        client_secret: process.env.GH_CLIENT_SECRET,
        state,
        code
      }
    });

    if (status === 200) {
      const qs = querystring.parse(data);
      if (qs.error) {
        return errorPage(res, 'Callback Error 3: ' + String(qs.error_description));
      } else {
        const port = Number(state.split('_')[1]);
        return redirectToUser(res, { token: qs.access_token }, port);
      }
    } else {
      return errorPage(res, 'Callback Error 4: GitHub server error');
    }
  }
  catch (err) {
    return errorPage(res, 'Callback Error 5: An uknown error occurred, maybe GitHub server is down right now');
  }
};



// https://micro-github.*USERNAME*.now.sh/api/login
// https://micro-github.*USERNAME*.now.sh/api/callback
// [name].ts https://vercel.com/docs/v2/serverless-functions/introduction#path-segments
module.exports = (req: NowRequest, res: NowResponse) => {
  switch (req.query.action) {
    case 'login':
      login(req, res); break;
    case 'callback':
      callback(req, res); break;
  }
};



function errorPage(res: NowResponse, errorMessage: string) {
  res.send(`
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <title>GitHub OAuth Error</title>
  </head>

  <body>
    <p>[ ${errorMessage} ]</p>

    <style>
      * {
        padding: 0;
        margin: 0;
      }
      p {
        margin-top: 14px;
        font-size: 18px;
        font-family: sans-serif;
        text-align: center;
      }
    </style>
  </body>
</html>

`);
}
